agent/command/git.go

```diff
@@ -393,240 +434,260 @@ func (c *gitFetchProject) Execute(ctx context.Context, comm client.Communicator,
+	logger.Execution().Infof("Fetching module: %s", moduleName)
+
+	var module *model.Module
+	module, err = conf.Project.GetModuleByName(moduleName)
+	if err != nil {
+		return errors.Errorf("Couldn't get module %s: %v", moduleName, err)
+	}
+	if module == nil {
+		return errors.Errorf("No module found for %s", moduleName)
 	}
 
-	var projectMethod string
-	var projectToken string
-	projectMethod, projectToken, err = getProjectMethodAndToken(c.Token, conf.Expansions.Get(evergreen.GlobalGitHubTokenExpansion), conf.Distro.CloneMethod)
+	moduleBase := filepath.ToSlash(filepath.Join(expandModulePrefix(conf, module.Name, module.Prefix, logger), module.Name))
+
+	var revision string
+	// use submodule revisions based on the main patch. If there is a need in the future,
+	// this could maybe use the most recent submodule revision of all requested patches.
+	// We ignore set-module changes for commit queue, since we should verify HEAD before merging.
+	if p != nil {
+		patchModule := p.FindModule(moduleName)
+		if patchModule != nil {
+			if conf.Task.Requester == evergreen.MergeTestRequester {
+				revision = module.Branch
+				c.logModuleRevision(logger, revision, moduleName, "defaulting to HEAD for merge")
+			} else {
+				revision = patchModule.Githash
+				if revision != "" {
+					c.logModuleRevision(logger, revision, moduleName, "specified in set-module")
+				}
+			}
+		}
+	}
+	if revision == "" {
+		revision = c.Revisions[moduleName]
+		if revision != "" {
+			c.logModuleRevision(logger, revision, moduleName, "specified as parameter to git.get_project")
+		}
+	}
+	if revision == "" {
+		revision = conf.Expansions.Get(moduleRevExpansionName(moduleName))
+		if revision != "" {
+			c.logModuleRevision(logger, revision, moduleName, "from manifest")
+		}
+	}
+	// if there is no revision, then use the revision from the module, then branch name
+	if revision == "" {
+		if module.Ref != "" {
+			revision = module.Ref
+			c.logModuleRevision(logger, revision, moduleName, "ref field in config file")
+		} else {
+			revision = module.Branch
+			c.logModuleRevision(logger, revision, moduleName, "branch field in config file")
+		}
+	}
+	var owner, repo string
+	owner, repo, err = thirdparty.ParseGitUrl(module.Repo)
 	if err != nil {
-		return errors.Wrap(err, "failed to get method of cloning and token")
+		return err
 	}
-	opts := cloneOpts{
-		method:             projectMethod,
-		owner:              conf.ProjectRef.Owner,
-		repo:               conf.ProjectRef.Repo,
-		branch:             conf.ProjectRef.Branch,
-		dir:                c.Directory,
-		token:              projectToken,
-		shallowClone:       c.ShallowClone && !conf.Distro.DisableShallowClone,
-		recurseSubmodules:  c.RecurseSubmodules,
-		mergeTestRequester: conf.Task.Requester == evergreen.MergeTestRequester,
+
+	var modulePatch *patch.ModulePatch
+	if p != nil {
+		// find module among the patch's Patches
+		for i := range p.Patches {
```

Comment:
We already do this at 508 with p.FindModule so i think we can just store that variable outside of the if and skip this altogether